File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONRESPONSE.cs
﻿using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_LOGONRESPONSE : Message
    {
        public enum Statuses : UInt32
        {
            Failure = 0,
            Success = 1,
        };

        public SID_LOGONRESPONSE()
        {
            Id = (byte)MessageIds.SID_LOGONRESPONSE;
            Buffer = new byte[0];
        }

        public SID_LOGONRESPONSE(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_LOGONRESPONSE;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            if (context == null || context.Client == null || !context.Client.Connected || context.Client.GameState == null) return false;

            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");

                        if (Buffer.Length < 29)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 29 bytes");

                        if (context.Client.GameState.ActiveAccount != null)
                            throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} cannot be sent after logging into an account");

                        /**
                         * (UINT32)     Client Token
                         * (UINT32)     Server Token
                         *  (UINT8)[20] Password Hash
                         * (STRING)     Username
                         */

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var clientToken = r.ReadUInt32();
                        var serverToken = r.ReadUInt32();
                        var passwordHash = r.ReadBytes(20);
                        context.Client.GameState.Username = r.ReadString();

                        if (!Battlenet.Common.AccountsDb.TryGetValue(context.Client.GameState.Username, out Account account) || account == null)
                        {
                            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] does not exist");
                            return new SID_LOGONRESPONSE().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.Failure }}));
                        }

                        var passwordHashDb = (byte[])account.Get(Account.PasswordKey, new byte[20]);
                        var compareHash = OldAuth.CheckDoubleHashData(passwordHashDb, clientToken, serverToken);
                        if (!compareHash.SequenceEqual(passwordHash))
                        {
                            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon failed password mismatch");
                            account.Set(Account.FailedLogonsKey, ((UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0)) + 1);
                            return new SID_LOGONRESPONSE().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.Failure }}));
                        }

                        var flags = (Account.Flags)account.Get(Account.FlagsKey, Account.Flags.None);
                        if ((flags & Account.Flags.Closed) != 0)
                        {
                            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon failed account closed");
                            account.Set(Account.FailedLogonsKey, ((UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0)) + 1);
                            return new SID_LOGONRESPONSE().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.Failure }}));
                        }

                        context.Client.GameState.ActiveAccount = account;
                        context.Client.GameState.FailedLogons = (UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0);
                        context.Client.GameState.LastLogon = (DateTime)account.Get(Account.LastLogonKey, DateTime.Now);

                        account.Set(Account.FailedLogonsKey, (UInt32)0);
                        account.Set(Account.IPAddressKey, context.Client.RemoteEndPoint.ToString().Split(":")[0]);
                        account.Set(Account.LastLogonKey, DateTime.Now);
                        account.Set(Account.PortKey, context.Client.RemoteEndPoint.ToString().Split(":")[1]);

                        var serial = 1;
                        var onlineName = context.Client.GameState.Username;
                        while (!Battlenet.Common.ActiveAccounts.TryAdd(onlineName, account)) onlineName = $"{context.Client.GameState.Username}#{++serial}";
                        context.Client.GameState.OnlineName = onlineName;

                        context.Client.GameState.Username = (string)account.Get(Account.UsernameKey, context.Client.GameState.Username);

                        if (!Battlenet.Common.ActiveGameStates.TryAdd(context.Client.GameState.OnlineName, context.Client.GameState))
                        {
                            Logging.WriteLine(Logging.LogLevel.Error, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Failed to add game state to active game state cache");
                            account.Set(Account.FailedLogonsKey, ((UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0)) + 1);
                            Battlenet.Common.ActiveAccounts.TryRemove(onlineName, out _);
                            return false;
                        }

                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon success as [{context.Client.GameState.OnlineName}]");
                        if (!(new SID_LOGONRESPONSE()).Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.Success }}))) return false;

                        var emailAddress = account.Get(Account.EmailKey, new byte[0]);
                        if (emailAddress.Length == 0)
                        {
                            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] does not have an email set");
                            new SID_SETEMAIL().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                        }

                        return true;
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Status
                         */

                        var status = (UInt32)(Statuses)context.Arguments["status"];

                        Buffer = new byte[4];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write(status);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes) (status: 0x{status:X8})");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_LOGONRESPONSE.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_LOGONRESPONSE
        Inherits Message

        Protected Enum Statuses As UInt32
            Failure = 0
            Success = 1
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_LOGONRESPONSE)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_LOGONRESPONSE)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Dim rereAccount As Account = Nothing

            Select Case context.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 29 Then Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be at least 29 bytes")

                    Dim clientToken, serverToken As UInt32
                    Dim passwordHash() As Byte
                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            clientToken = r.ReadUInt32()
                            serverToken = r.ReadUInt32()
                            passwordHash = r.ReadBytes(20)
                            context.Client.GameState.Username = r.ReadString()
                        End Using
                    End Using

                    Battlenet.Common.AccountsDb.TryGetValue(context.Client.GameState.Username, rereAccount)

                    If rereAccount Is Nothing Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] does not exist")
                        Return New SID_LOGONRESPONSE().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.Failure}
                        }))
                    End If

                    Dim passwordHashDb = CType(rereAccount.[Get](Account.PasswordKey, New Byte(19) {},), Byte())
                    Dim compareHash = OldAuth.CheckDoubleHashData(passwordHashDb, clientToken, serverToken)

                    If Not compareHash.SequenceEqual(passwordHash) Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon failed password mismatch")
                        rereAccount.[Set](Account.FailedLogonsKey, (CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))) + 1)
                        Return New SID_LOGONRESPONSE().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.Failure}
                        }))
                    End If

                    Dim flags = CType(rereAccount.[Get](Account.FlagsKey, Account.Flags.None,), Account.Flags)

                    If (flags And Account.Flags.Closed) <> 0 Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon failed account closed")
                        rereAccount.[Set](Account.FailedLogonsKey, (CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))) + 1)
                        Return New SID_LOGONRESPONSE().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.Failure}
                        }))
                    End If

                    context.Client.GameState.ActiveAccount = rereAccount
                    context.Client.GameState.FailedLogons = CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))
                    context.Client.GameState.LastLogon = CDate(rereAccount.[Get](Account.LastLogonKey, DateTime.Now,))
                    rereAccount.[Set](Account.FailedLogonsKey, CUInt(0))
                    rereAccount.[Set](Account.IPAddressKey, context.Client.RemoteEndPoint.ToString().Split(":")(0))
                    rereAccount.[Set](Account.LastLogonKey, DateTime.Now)
                    rereAccount.[Set](Account.PortKey, context.Client.RemoteEndPoint.ToString().Split(":")(1))

                    SyncLock Battlenet.Common.ActiveAccounts
                        Dim serial = 1
                        Dim onlineName = context.Client.GameState.Username

                        While Battlenet.Common.ActiveAccounts.ContainsKey(onlineName)
                            onlineName = $"{context.Client.GameState.Username}#{System.Threading.Interlocked.Increment(serial)}"
                        End While

                        context.Client.GameState.OnlineName = onlineName
                        Battlenet.Common.ActiveAccounts.Add(onlineName, rereAccount)
                    End SyncLock

                    context.Client.GameState.Username = CStr(rereAccount.[Get](Account.UsernameKey, context.Client.GameState.Username))

                    SyncLock Battlenet.Common.ActiveGameStates
                        Battlenet.Common.ActiveGameStates.Add(context.Client.GameState.OnlineName, context.Client.GameState)
                    End SyncLock

                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"Account [{context.Client.GameState.Username}] logon success as [{context.Client.GameState.OnlineName}]")
                    Return New SID_LOGONRESPONSE().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                        {"status", Statuses.Success}
                    }))
                Case MessageDirection.ServerToClient
                    Dim status = CUInt(CType(context.Arguments("status"), Statuses))
                    Buffer = New Byte(3) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(status)
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes) (status: 0x{status})")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
